core/server/src/main/java/alluxio/worker/block/evictor/LIRSEvictor.java

```diff
@@ -0,0 +1,573 @@
+/**
+ * This class is used to evict blocks with LIRS policy on each tier. Different from LRU and LRFU,
+ * LIRS evicts the block with maximum Inter-Reference Recency (IRR). IRR is the number of distinct
+ * blocks between last access and second-to-last access to the block. Block with high IRR is
+ * predicted to be accessed again in a long time, so they need to be evicted first to make room for
+ * other blocks which will be accessed soon. LIRS divides the tier into 2 parts: LIR cache and HIR
+ * cache. LIR cache holds blocks with low IRR and HIR cache holds blocks with high IRR. So blocks
+ * in HIR cache needs to be evicted first. If blocks in HIR cache gets a low IRR, they will be
+ * moved to LIR cache. And if the LIR cache is full, some blocks with relatively high IRR will be
+ * moved to HIR cache. LIRS can achieve better performance than LRU and LRFU in loop access pattern.
+ * For example, a list of files contain blocks: a1,a2,...,ak. When iterating accessing these files,
+ * the blocks will be referenced in the order: a1,a2,...,ak,a1,a2,...,ak,.... some blocks of them
+ * will be stored in LIR cache and the others in HIR cache. No HIR blocks will be moved to LIR cache
+ * because all blocks have equal IRR. Therefore, all blocks in LIR cache will never be removed which
+ * promise the hit rate is at least (number of LIR blocks)/(number of all blocks). while if these
+ * blocks cannot be held in the memory in this case, LRU will achieve zero hit rate. It's also
+ * verified with Spark kmeans that the hit rate of LIRS policy improves a lot than other policies.
+ * So if your workloads contain a lot of iterations like kmeans, LIRS policy is recommended for you.
+ * Besides, LIRS can achieve as good performance as LRU and LRFU in most of workloads.
+ */
+public final class LIRSEvictor extends AbstractEvictor {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final int LINKED_HASH_MAP_INIT_CAPACITY = 200;
+  private static final float LINKED_HASH_MAP_INIT_LOAD_FACTOR = 0.75f;
+  private static final boolean LINKED_HASH_MAP_ACCESS_ORDERED = false;
+  private static final boolean UNUSED_MAP_VALUE = true;
+
+  // The percent of HIR blocks on each tier
+  private final double mHIRPercent;
+  // The percent of LIR blocks on each tier
+  private final double mLIRPercent;
+  private final Configuration mConfiguration;
+  // Map from the location of a StorageDir to its condition of used space
+  private Map<BlockStoreLocation, SpaceContainer> mSpaceManager =
+      new ConcurrentHashMap<BlockStoreLocation, SpaceContainer>();
+  /**
+   * Map from the the block id to the block size. Maintain the information of block size is because
+   * the space of HIR block or LIR block need to be reclaimed when
+   * {@link #onRemoveBlockByClient(long, long)} or {@link #onRemoveBlockByWorker(long, long)} is
+   * called.
+   */
+  private Map<Long, Long> mBlockIdToSize = new ConcurrentHashMap<Long, Long>();

```

Comment:
@Yours-shupeng Other event listeners may need block's info in `onRemoveByXXX` too, it may be reasonable to change their definition into `void onRemoveBlockByXXX(long sessionId, BlockMeta blockMeta);`, then event listeners do not need to maintain similar information caches. What do you think?